### PR TITLE
DCC(nuke) context switch: verify that new shot/entity is Rez-compatible with current (RND-342)

### PR DESCRIPTION
**Pull request by gst**
# Purpose #

Allow change of context when under Rez and detects if the newly desired context (shot or entity) is compatible with the current/previous one.
If not launch a new DCC via rdo_rez_launcher regularly.

# Risk # 

Should be low : 
+ when not under rez this is a no-op basically.
+ when under rez: well the check is done and if need to be a new DCC process instance is started (the current one is kept open).

# Requires #

https://bitbucket.org/rodeofx/rdo_rez_launcher/pull-requests/24/rnd-342-nuke-context-switch-verify-that/diff

### CODE REVIEW

File: python/tk_multi_workfiles/work_files.py [hunk 1]
```diff
@@ -1050,3 +1058,66 @@ def _get_app_settings_for_context(self, context):
...
+        return True
+
+    sorted_prev_rez_used_request = sorted(rez_used_request.split())
+
+    # get the "rodeo facility" project :
+    rodeo_facility = new_ctx.sgtk.shotgun.find_one(
+        'CustomNonProjectEntity02', [['code', 'is', 'All rodeo']], fields=['code'])
+
+    # delayed import on purpose, because rdo_rez_launcher isn't available outside of rez ..
+    import rdo_rez_launcher.workflow
```
Comment (JeanChristopheMorinRodeoFX) [inline python/tk_multi_workfiles/work_files.py]: It is not a guaranty that it will be available. Tank and its apps can be used without any DCC/launcher (which I personally do from time to time). If `rdo_rez_launcher` is not available, no check should be done.

**Comment by JeanChristopheMorinRodeoFX** [comment]
As said in private, it's not the only app used to change a context. We will also need to handle them.

The `tk-multi-workfiles2` is also being used in prod. Katana is not even using any toolkit apps, they are using this: https://bitbucket.org/rodeofx/tk-multi-contextswitcher/src/87422efbf50b2b18d3a947bc160982966344ec57/python/contextSwitcher/contextSwitcher.py?at=dev&fileviewer=file-view-default#contextSwitcher.py-11

Note that `tk-multi-workfiles2` is using the `change_context` (new way), while as you might have seen `tk-multi-workfiles1` is using the `destroy/start` way (old way).

**Comment by gst** [comment]
> As said in private, 
> tk-multi-workfiles2 

this will obviously be done in a separate PR ;)
but thanks for remind I'll create a new Jira issue/task to continue on this.

File: python/tk_multi_workfiles/work_files.py [hunk 2]
```diff
@@ -72,7 +73,7 @@ def __init__(self, app):
         self._work_area_template = None
         self._publish_template = None
         self._publish_area_template = None
-        
+
```
Comment (JeanChristopheMorinRodeoFX) [inline python/tk_multi_workfiles/work_files.py]: SInce this is not our app, I would avoid doing not needed changes as this. Updating Tank apps is already painful sometimes, so less changes is the best. Thanks

File: python/tk_multi_workfiles/work_files.py [hunk 3]
```diff
@@ -779,8 +789,6 @@ def change_work_area(self, enable_start_new=True):
             
             if self._do_change_work_area(context_and_flags[1]):
                 break
-
```
Comment (JeanChristopheMorinRodeoFX) [inline python/tk_multi_workfiles/work_files.py]: Same here, unnecessary changes.

File: python/tk_multi_workfiles/work_files.py [hunk 4]
```diff
@@ -779,8 +789,6 @@ def change_work_area(self, enable_start_new=True):
             
             if self._do_change_work_area(context_and_flags[1]):
                 break
-            
-
```
Comment (JeanChristopheMorinRodeoFX) [inline python/tk_multi_workfiles/work_files.py]: And here.

File: python/tk_multi_workfiles/work_files.py [hunk 5]
```diff
@@ -819,7 +827,7 @@ def _update_current_work_area(self, ctx):
         if ctx == self._context:
             self._app.log_debug("Context hasn't changed so nothing to do!")    
             return
-        
+
```
Comment (JeanChristopheMorinRodeoFX) [inline python/tk_multi_workfiles/work_files.py]: And here again.

File: python/tk_multi_workfiles/work_files.py [hunk 6]
```diff
@@ -1050,3 +1058,66 @@ def _get_app_settings_for_context(self, context):
...
+    res = QtGui.QMessageBox.question(
+        None, "Rez incompatible contexts ..", msg,
+        QtGui.QMessageBox.Yes | QtGui.QMessageBox.No)
+
+    if res == QtGui.QMessageBox.Yes:
+        cmd_args = [
+            os.environ['RDO_REZ_LAUNCHER_PATH'],  # given by rdo_rez_launcher itself..
+            cur_show,
+            new_ctx.entity['name'],
```
Comment (JeanChristopheMorinRodeoFX) [inline python/tk_multi_workfiles/work_files.py]: `entity` is not guaranteed to be there. The use case is small, but it could happen.

File: python/tk_multi_workfiles/work_files.py [hunk 7]
```diff
@@ -1050,3 +1058,66 @@ def _get_app_settings_for_context(self, context):
...
+    rez_used_request = os.getenv('REZ_USED_REQUEST')
+    if not rez_used_request:
+        # not under Rez so it's always "ok"..
+        return True
+
+    sorted_prev_rez_used_request = sorted(rez_used_request.split())
+
+    # get the "rodeo facility" project :
+    rodeo_facility = new_ctx.sgtk.shotgun.find_one(
+        'CustomNonProjectEntity02', [['code', 'is', 'All rodeo']], fields=['code'])
```
Comment (JeanChristopheMorinRodeoFX) [inline python/tk_multi_workfiles/work_files.py]: Not the first time I say it: `CustomNonProjectEntity02` should be defined in a constant. Possibly `rdo_shotgun_workflows.constants` if this function get away from this current file. Or maybe in `rdo_rez_launcher`.

Comment (JeanChristopheMorinRodeoFX) [reply python/tk_multi_workfiles/work_files.py]: And if I'm not mistaken `All rodeo` is already present in a constant. If not, then it should be added as a constant in `rdo_rez_launcher`

**Review by JeanChristopheMorinRodeoFX — CHANGES_REQUESTED** [review]

File: python/tk_multi_workfiles/work_files.py (hunk 2 above)
Comment (gst) [reply python/tk_multi_workfiles/work_files.py]: yeah I agree ; I shouldn't have committed that. I had firstly done some other change here and there but my pycharm has an auto-del-space-from-empty-lines and anyway I shouldn't have commit that.
I'll fix.

File: python/tk_multi_workfiles/work_files.py (hunk 7 above)
Comment (gst) [reply python/tk_multi_workfiles/work_files.py]: well seen..

File: python/tk_multi_workfiles/work_files.py (hunk 2 above)
Comment (JeanChristopheMorinRodeoFX) [reply python/tk_multi_workfiles/work_files.py]: Great!

File: python/tk_multi_workfiles/work_files.py (hunk 7 above)
Comment (gst) [reply python/tk_multi_workfiles/work_files.py]: heuh for CustomNonProjectEntity02 I won't make rdo_rez_launcher dependent on rdo_shotgun_workflows. not a chance.

File: python/tk_multi_workfiles/work_files.py (hunk 1 above)
Comment (gst) [reply python/tk_multi_workfiles/work_files.py]: if under Rez it should(will) be present..

File: python/tk_multi_workfiles/work_files.py (hunk 6 above)
Comment (gst) [reply python/tk_multi_workfiles/work_files.py]: If they have selected an entity (shot or asset) then it's there. If they don't (cancel or I don't know) : then this code path isn't called - from what I could see..

File: python/tk_multi_workfiles/work_files.py (hunk 1 above)
Comment (JeanChristopheMorinRodeoFX) [reply python/tk_multi_workfiles/work_files.py]: The thing is, Tank can be used outside of Rez. Tank and its apps/engines should continue to work outside of REZ and any launcher.

File: python/tk_multi_workfiles/work_files.py (hunk 7 above)
Comment (gst) [reply python/tk_multi_workfiles/work_files.py]: `All rodeo` it is actually defined in the config file used by rdo_rez_launcher...

but I agree it could be made more constants because I don't see why we would change it..  (it could break lot of stuffs I can guess).

hmm.. ok let's put it as const in our rez launcher..

File: python/tk_multi_workfiles/work_files.py (hunk 1 above)
Comment (gst) [reply python/tk_multi_workfiles/work_files.py]: I know but 2-4 lines above there is a check about REZ_USED_REQUEST  AND even commented : that is if we reach the import line : we are under Rez...

File: python/tk_multi_workfiles/work_files.py [hunk 8]
```diff
@@ -754,6 +754,23 @@ def select_work_area(self, mode=SelectWorkAreaForm.SELECT_WORK_AREA):
...
+                # delayed import on purpose, not available outside of rez:
+                new_ctx = widget.context
+                from rdo_rez_launcher.tools import start_app_if_new_entity_incompatible
+                if start_app_if_new_entity_incompatible(
+                    lambda title, message: QtGui.QMessageBox.Yes == QtGui.QMessageBox.question(
+                        None, title, message, QtGui.QMessageBox.Yes | QtGui.QMessageBox.No),
+                    new_ctx.tank.shotgun, new_ctx.entity['name'],
+                    tank_config_name=new_ctx.tank.configuration_name
+                ):
+                    # this means the new entity isn't compatible with current one and we stop here,
```
Comment (darkvertex) [inline python/tk_multi_workfiles/work_files.py:769]: Had to ask you what you meant by "compatible". Apparently what you mean is that you check for rez package conflicts since there could be different packages that can clash between different entities.

That makes sense but I think this line could be a bit more explanatory. ;)

**Review by darkvertex — CHANGES_REQUESTED** [review]
Code LGTM but I'd like a clearer comment about what "incompatible" means. Explain that rez package conflicts can happen because different entities can have completely different packages, maybe?

**Review by darkvertex — APPROVED** [review]
Much better, thanks! :+1:
